### ISSUE TITLE
Bad interaction of type name shadowing and printing type names

### ISSUE THREAD

**Issue by frank-emrich** [links-lang/links#594]
Links currently allows shadowing of type names. Consider the following program:

```
typename Foo = Int;

sig x : Foo
var x = 3;

typename Foo = String;

sig y : Foo
var y = x;
```

This program is ill-typed because we try to assign an Int to a variable that is supposed to be a String. However, the error message is comically bad:

```
../my_examples/typename_shadowing.links:9: Type error: The value has type
    `Foo'
but it is annotated with type
    `Foo'
In expression: var y = x.
```

What happened? When translating type names to semantic types, we handle scoping correctly. However, when we annotate an AST node with a named type, we create an `Alias` node. It contains the (correct) semantic type, but also a string representation of the type name.
This means that the type of x is something like `Alias ("Foo", Int)` and the expected type of y is `Alias ("Foo", String)`. When printing a type that is an `Alias` node, we currently always print the string name, even if that name is not valid (e.g., due to shadowing) anymore.

There are multiple solutions:
1. Ignore this as it is only a printing issue. The semantic types are correct.
2. Disallow shadowing of type names
3. Try to be more clever about printing types. When seeing an alias of the form `Alias (name, def)`, we could check if `name` in the *current* type environment is the same as `def`. In this case, we just print `name`. If it's not, we don't know how to print this type alias. We could either print its unrolling (i.e., print def) or we could try to be more clever by printing something like "Foo (as defined on line x in file y)".

**Comment by jstolarek** [comment]
This echoes something I already said when discussing mutually recursive definitions (https://github.com/links-lang/links/issues/457#issuecomment-471626302) - I see the ability to redefine top-level definitions as a misfeature. Same here. Therefore, (2) Disallow shadowing of type names.

Regarding (3), see #522

**Comment by jamescheney** [comment]
Do the types have enough location information that we can disambiguate by printing the declaration locations? E.g.

```
../my_examples/typename_shadowing.links:9: Type error: The value has type
    `Foo' (defined at : typename_shadowing.links:12)
but it is annotated with type
    `Foo' (defined at : typename_shadowing.links:34)
In expression: var y = x.
```

This could be helpful if the two `Foo`s are defined in different modules and shadow due to `open`.

**Comment by jamescheney** [comment]
Which could happen even if we adopt Janek's  suggestion of disallowing lexical shadowing.

**Comment by frank-emrich** [comment]
> Do the types have enough location information that we can disambiguate by printing the declaration locations?


Currently not, but that shouldn't be too difficult to implement. It just adds a dependency from the types to the lexing information (module SourceCode), which is a bit weird.



The University of Edinburgh is a charitable body, registered in Scotland, with registration number SC005336.

**Comment by slindley** [comment]
I disagree with @jstolarek on shadowing. There are certainly pros and cons of allowing and disallowing shadowing at the top level, but I would regard shadowing as the canonical behaviour for ML-like languages, which Links currently is. I would only consider changing this behaviour if there were a compelling reason to do so, which I have not seen.

The usual reason for allowing shadowing is uniformity / compositionality. A concrete side-effect which can be of benefit is that sometimes the programmer wants to hide a variable within a particular scope and shadowing provides a means to do so. One reason people complain about shadowing is confusion over which definition a name refers to (of course it always refers to the closest lexically enclosing binding). The reason Haskell doesn't allow shadowing is because Haskell takes advantage of having limited effects by default to avoid the need to explicitly delimit mutual recursion, a feature which is incompatible with shadowing. But in strict ML-like side-effecting languages like Links, that's pretty much a non-issue because mutual recursion must be delimited somehow (whether implicitly by let bindings as we used to do or explicitly as we do now).

I do like @jamescheney's suggestion. The thing to watch out for in practice is that it isn't necessarily clear when the user would like to see the extra meta data. A starting point could be to make sure we always elaborate when two different types with the same name appear in the same error message. It would also be nice if we could generate HTML error messages in the web browser where the link to the definition site is attached to the typename even if it isn't necessarily visible in the message.

Of course this problem is not remotely new. Most implementations of ML suffer with it in one way or another. I remember Dave Matthews (the main Poly/ML developer) talking about exactly this problem about 10 years ago... I can't remember what solution he proposed, though.

**Comment by jamescheney** [comment]
I guess this is more of an annoyance because at the moment the same types are used in the source/sugared AST as well as the IR, which means the type information could persist to the IR stage where it will never be needed.  I don't see any other particular problem with this, and anyway the right thing to do (eventually) is to typecheck the IR and then erase all the type information, I think.

I also think it could be interesting/worthwhile to keep track of how an identifier got to be in scope (e.g. via open/import) so that for say something like this:

```
module A { 
  typename Foo = Int;
  
  sig x : Foo
  var x = 3;
}

module B { 
  typename Foo = String;
}

open A
open B
sig y : Foo
var y = x;
```
the error message would say something like 
```
../my_examples/typename_shadowing.links:9: Type error: The value has type
    `Foo' (defined at : typename_shadowing.links:2, from module A opened at typename_shadowing.links:12)
but it is annotated with type
    `Foo' (defined at : typename_shadowing.links:9, from module B opened at typename_shadowing.links:13)
In expression: var y = x.
```

But that's because whenever I see a bad error message nowadays I assume that more provenance would make it better, which is not necessarily the case.

(Note also that this is a concrete example that `open` can cause shadowing even if we don't allow repeated definitions of `typename Foo` in the same lexical scope.)

**Comment by jamescheney** [comment]
Yes, most MLs have terrible error messages when you have multiple modules defining a type with the same name e.g. `t` (including sharing / where type constraints that imply some type equalities)...

I agree that the metadata could become overwhelming (think of multiple open chains, functors, etc.)

+1 to having Links generate HTML error messages that one can interact with (we should record this as a separate issue).   Then the "explanation" metadata could be hyperlinks / highlighting instead of natural language text.  These would also be super-useful for TryLinks.

**Comment by frank-emrich** [comment]
@slindley 
>  A starting point could be to make sure we always elaborate when two different types with the same name appear in the same error message. 

That's not sufficient. Consider a slightly modified example:
```
typename Foo = Int

sig x : Foo
var x = 3;

typename Foo = String;
x
```
This program will print `3 : Foo`.

The idea I mention in my original post was the following: When printing a type `Alias (name, st)`, first check if looking up `name` in the current scope yields a type equal to `st`. If so, just print `name`. If not, elaborate by either printing the source location in addition to the name, or just its definition, omitting the name.

@jamescheney 
> I also think it could be interesting/worthwhile to keep track of how an identifier got to be in scope (e.g. via open/import) so that for say something like this

The module system will do some of that: It stores in the typing environment whether or not a binding came into scope due to an `open`, and if so, which module. Currently, it doesn't track source locations.

> Note also that this is a concrete example that open can cause shadowing even if we don't allow repeated definitions of typename Foo in the same lexical scope.

Indeed, I stumbled over these issues while writing tests for the module system. Fortunately, the scoping of the actual (semantic) types works correctly. It's just the printing that's broken.

**Comment by jstolarek** [comment]
> whenever I see a bad error message nowadays I assume that more provenance would make it better, which is not necessarily the case.

Additional meta-data could be hidden behind a flag, in the same way stack traces can be hidden by default.
